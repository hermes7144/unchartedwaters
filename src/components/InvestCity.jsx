import React, { useState } from 'react';
import { useQuery } from '@tanstack/react-query';
import { useInfoContext } from './context/InfoContext';
import { useInvestContext } from './context/InvestContext';

export default function InvestCity() {
  const { getAreas, getCitys, getGoods } = useInfoContext();
  const { target, current } = useInvestContext();
  const { isLoading: isLoadingAreas, data: areas } = useQuery(['areas'], getAreas);
  const { isLoading: isLoadingCitys, data: citys } = useQuery(['citys'], getCitys);
  const [selectedArea, setSelectedArea] = useState();
  const [selectedCity, setSelectedCity] = useState('');

  const { data: city } = useQuery(['city', selectedCity], () => getGoods(selectedCity), { staleTime: Infinity, enabled: !!selectedCity });
  const handleAreas = (e) => {
    setSelectedArea(e.target.value);
    setSelectedCity('');
  };

  const handleCitys = (e) => setSelectedCity(e.target.value);

  function setPrice(city) {
    switch (setClass(city)) {
      case 1:
        return parseFloat(Math.round((city.goods_price * 1.5).toFixed(1)));
      case 2:
        return parseFloat(Math.round((city.goods_price * 1.3).toFixed(1)));
      case 3:
        return parseFloat(Math.round((city.goods_price * 1.2).toFixed(1)));
      case 4:
        return parseFloat(Math.round((city.goods_price * 1.1).toFixed(1)));
      default:
        return city.goods_price;
    }
  }

  function setClass(city) {
    if (city.goods_highest_point + city.goods_dev_point <= target) return 1;
    else if (city.goods_highest_point * 0.8 + city.goods_dev_point <= target) return 2;
    else if (city.goods_highest_point * 0.6 + city.goods_dev_point <= target) return 3;
    else if (city.goods_highest_point * 0.4 + city.goods_dev_point <= target) return 4;
    else return 5;
  }

  const filtered = getFilteredCitys(citys, selectedArea);
  return (
    <article className='flex flex-col basis-1/2'>
      <h2 className='text-2xl font-bold my-4 text-center'>select</h2>
      <form className='flex flex-col px-12'>
        {(isLoadingAreas || isLoadingCitys) && <p>Loading...</p>}
        {areas && citys && (
          <>
            <label className=' font-bold' htmlFor='areas'>
              해역
            </label>
            <select id='areas' className='p-2 m-4 flex-1 border-2 border-dashed border-brand outline-none' onChange={handleAreas} value={selectedArea}>
              <option value=''>전체</option>
              {areas && areas.map((area, index) => <option key={index}>{area.area_nm}</option>)}
            </select>
            <label className=' font-bold' htmlFor='areas'>
              도시
            </label>
            <select id='areas' className='p-2 m-4 flex-1 border-2 border-dashed border-brand outline-none' onChange={handleCitys} value={selectedCity}>
              <option value=''>전체</option>
              {filtered.map((city, index) => (
                <option key={index}>{city.city_nm}</option>
              ))}
            </select>
          </>
        )}
      </form>

      <table className='border-collapse border border-slate-400 w-5/6 mx-auto'>
        <thead>
          <tr>
            <th className='border border-slate-300 p-4'>교역품</th>
            <th className='border border-slate-300 p-4'>출현 발전도</th>
            <th className='border border-slate-300 p-4'>가격</th>
            <th className='border border-slate-300 p-4'>수량</th>
            <th className='border border-slate-300 p-4'>등급</th>
            <th className='border border-slate-300 p-4'>원산물</th>
          </tr>
        </thead>
        <tbody>
          {city && city.length > 0 ? (
            city.map((city, index) => (
              <tr key={index}>
                <td className='border border-slate-300'>{city.goods_nm}</td>
                <td className={city.goods_dev_point > current && city.goods_dev_point <= target ? 'text-red-500 border border-slate-300 text-center' : city.goods_dev_point > current && city.goods_dev_point > target ? 'text-gray-300 border border-slate-300 text-center' : 'border border-slate-300 text-center'}>{city.goods_dev_point}</td>
                <td className='border border-slate-300 p-1 text-center'>{setPrice(city)}</td>
                <td className='border border-slate-300 p-1 text-center'>{city.goods_qty}</td>
                <td className='border border-slate-300 p-1 text-center'>{setClass(city)}</td>
                <td className='border border-slate-300 p-1 text-center'>{city.goods_add ? 'O' : 'X'}</td>
              </tr>
            ))
          ) : (
            <tr>
              <td className='text-center' colSpan='5'>
                도시를 선택해주세요.
              </td>
            </tr>
          )}
        </tbody>
      </table>
    </article>
  );
}
function getFilteredCitys(citys, filter) {
  if (filter) {
    return citys.filter((city) => city.city_area === filter);
  } else {
    return citys;
  }
}
