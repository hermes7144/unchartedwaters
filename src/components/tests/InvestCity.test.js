import { Route } from 'react-router-dom'
import { withAllContexts, withInvest, withRouter } from '../../tests/utils'
import InvestCity from '../InvestCity'
import { fakeAreas } from '../../tests/fakeAreas'
import { fakeCitys } from '../../tests/fakeCitys'
import { fakeGoods } from '../../tests/fakeGoods'
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

describe('InvestCity', () => {
  const getAreas = jest.fn();
  const getCitys = jest.fn();
  const getGoods = jest.fn();

  afterEach(() => {
    getAreas.mockReset();
    getCitys.mockReset();
    getGoods.mockReset();
  })

  it('renders correctly', async () => {
    getAreas.mockImplementation(() => fakeAreas)
    getCitys.mockImplementation(() => fakeCitys)
    getGoods.mockImplementation(() => fakeGoods)

    const { asFragment } = renderInvestCity();
    await waitForElementToBeRemoved(screen.queryByText('Loading...'))
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders select box correctly', async () => {
    getAreas.mockImplementation(() => fakeAreas)
    getCitys.mockImplementation(() => fakeCitys)
    getGoods.mockImplementation(() => fakeGoods)

    renderInvestCity();
    await waitForElementToBeRemoved(screen.queryByText('Loading...'))

    userEvent.selectOptions(
      // Find the select element
      screen.getByLabelText('해역'),
      // Find and select the Ireland option
      screen.getByRole('option', { name: '북해' }),
    )
    expect(screen.getByRole('option', { name: '북해' }).selected).toBe(true)
  })

  it('renders city goods', async () => {
    getAreas.mockImplementation(() => fakeAreas)
    getCitys.mockImplementation(() => fakeCitys)
    getGoods.mockImplementation(() => fakeGoods)

    renderInvestCity();
    await waitForElementToBeRemoved(screen.queryByText('Loading...'))


    const selectElement = screen.getByLabelText('도시');
    userEvent.selectOptions(selectElement, '리스본');
    const selectedOption = screen.getByRole('option', { name: '리스본' });
    expect(selectedOption.selected).toBeTruthy();
  })

  it('renders city goods snapshot', async () => {
    getAreas.mockImplementation(() => fakeAreas)
    getCitys.mockImplementation(() => fakeCitys)
    getGoods.mockImplementation(() => fakeGoods)

    const { asFragment } = renderInvestCity();
    await waitForElementToBeRemoved(screen.queryByText('Loading...'))

    const selectElement = screen.getByLabelText('도시');
    userEvent.selectOptions(selectElement, '리스본');

    expect(asFragment()).toMatchSnapshot();

  })


  function renderInvestCity() {
    return render(
      withAllContexts(
        withInvest(withRouter(<Route path='/' element=
          {< InvestCity />} />)),
        getAreas,
        getCitys,
        getGoods
      ),
    )
  }
})

