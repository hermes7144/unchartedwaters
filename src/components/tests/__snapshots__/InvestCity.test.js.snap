// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestCity renders city goods snapshot 1`] = `
<DocumentFragment>
  <article
    class="flex flex-col basis-1/2"
  >
    <h2
      class="text-2xl font-bold my-4 text-center"
    >
      select
    </h2>
    <form
      class="flex flex-col px-12"
    >
      <label
        class="font-bold"
        for="areas"
      >
        해역
      </label>
      <select
        class="p-2 m-4 flex-1 border-2 border-dashed border-brand outline-none"
        id="areas"
      >
        <option
          value=""
        >
          전체
        </option>
        <option>
          북해
        </option>
        <option>
          지중해
        </option>
      </select>
      <label
        class="font-bold"
        for="citys"
      >
        도시
      </label>
      <select
        class="p-2 m-4 flex-1 border-2 border-dashed border-brand outline-none"
        id="citys"
      >
        <option
          value=""
        >
          전체
        </option>
        <option>
          런던
        </option>
        <option>
          암스테르담
        </option>
        <option>
          리스본
        </option>
      </select>
    </form>
    <table
      class="border-collapse border border-slate-400 w-5/6 mx-auto"
    >
      <thead>
        <tr>
          <th
            class="border border-slate-300 p-4"
          >
            교역품
          </th>
          <th
            class="border border-slate-300 p-4"
          >
            출현 발전도
          </th>
          <th
            class="border border-slate-300 p-4"
          >
            가격
          </th>
          <th
            class="border border-slate-300 p-4"
          >
            수량
          </th>
          <th
            class="border border-slate-300 p-4"
          >
            등급
          </th>
          <th
            class="border border-slate-300 p-4"
          >
            원산물
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="text-center"
            colspan="5"
          >
            도시를 선택해주세요.
          </td>
        </tr>
      </tbody>
    </table>
  </article>
</DocumentFragment>
`;